File: packages/mcp-protocol/src/policy-engine.ts
import type { AgentPolicy, PolicyRules } from '@authbox/shared';

export interface AccessRequest {
  agentId: string;
  action: 'read' | 'use' | 'proxy';
  itemType?: string;
  itemId?: string;
  folderId?: string;
}

export interface AccessDecision {
  allowed: boolean;
  reason: string;
  appliedPolicies: string[];
  /** When step-up is required, this contains a request ID to track the approval. */
  pendingApprovalId?: string;
}

export interface PendingApproval {
  id: string;
  agentId: string;
  action: string;
  itemType?: string;
  itemId?: string;
  reason: string;
  createdAt: number;
  resolve: (approved: boolean) => void;
}

export class PolicyEngine {
  private rateLimitCounters = new Map<string, { count: number; windowStart: number }>();
  private pendingApprovals = new Map<string, PendingApproval>();

  /** Callback invoked when step-up approval is needed. Set by the MCP server. */
  onApprovalNeeded?: (approval: PendingApproval) => void;

  evaluate(policies: AgentPolicy[], request: AccessRequest): AccessDecision {
    if (policies.length === 0) {
      return { allowed: false, reason: 'No policies defined', appliedPolicies: [] };
    }

    const enabled = policies
      .filter((p) => p.enabled)
      .sort((a, b) => b.priority - a.priority);

    if (enabled.length === 0) {
      return { allowed: false, reason: 'All policies disabled', appliedPolicies: [] };
    }

    const applied: string[] = [];

    for (const policy of enabled) {
      const result = this.evaluatePolicy(policy, request);
      applied.push(policy.id);

      if (!result.allowed) {
        return { allowed: false, reason: result.reason, appliedPolicies: applied };
      }
    }

    return { allowed: true, reason: 'All policies passed', appliedPolicies: applied };
  }

  private evaluatePolicy(
    policy: AgentPolicy,
    request: AccessRequest,
  ): { allowed: boolean; reason: string } {
    const rules = policy.rules;

    switch (policy.policyType) {
      case 'item_scope':
        return this.checkItemScope(rules, request);
      case 'action_perm':
        return this.checkActionPermission(rules, request);
      case 'rate_limit':
        return this.checkRateLimit(request.agentId, rules);
      case 'time_window':
        return this.checkTimeWindow(rules);
      case 'step_up':
        return this.checkStepUp(rules, request);
      default:
        return { allowed: true, reason: 'Unknown policy type, defaulting to allow' };
    }
  }

  private checkItemScope(
    rules: PolicyRules,
    request: AccessRequest,
  ): { allowed: boolean; reason: string } {
    if (rules.deniedItemIds?.includes(request.itemId ?? '')) {
      return { allowed: false, reason: 'Item explicitly denied' };
    }

    if (rules.allowedItemTypes && request.itemType) {
      const allowed = rules.allowedItemTypes as string[];
      if (!allowed.includes(request.itemType)) {
        return { allowed: false, reason: `Item type '${request.itemType}' not in allowed types` };
      }
    }

    if (rules.allowedItemIds && request.itemId) {
      if (!rules.allowedItemIds.includes(request.itemId)) {
        return { allowed: false, reason: 'Item ID not in allowed list' };
      }
    }

    if (rules.allowedFolderIds && request.folderId) {
      if (!rules.allowedFolderIds.includes(request.folderId)) {
        return { allowed: false, reason: 'Folder not in allowed list' };
      }
    }

    return { allowed: true, reason: 'Item scope check passed' };
  }

  private checkActionPermission(
    rules: PolicyRules,
    request: AccessRequest,
  ): { allowed: boolean; reason: string } {
    if (!rules.allowedActions || rules.allowedActions.length === 0) {
      return { allowed: true, reason: 'No action restrictions' };
    }

    if (!rules.allowedActions.includes(request.action)) {
      return { allowed: false, reason: `Action '${request.action}' not permitted` };
    }

    return { allowed: true, reason: 'Action permitted' };
  }

  private checkRateLimit(
    agentId: string,
    rules: PolicyRules,
  ): { allowed: boolean; reason: string } {
    if (!rules.maxRequests || !rules.windowSeconds) {
      return { allowed: true, reason: 'No rate limit configured' };
    }

    const now = Date.now();
    const windowMs = rules.windowSeconds * 1000;
    const key = agentId;
    const counter = this.rateLimitCounters.get(key);

    if (!counter || now - counter.windowStart > windowMs) {
      this.rateLimitCounters.set(key, { count: 1, windowStart: now });
      return { allowed: true, reason: 'Rate limit check passed' };
    }

    if (counter.count >= rules.maxRequests) {
      return { allowed: false, reason: `Rate limit exceeded: ${rules.maxRequests} requests per ${rules.windowSeconds}s` };
    }

    counter.count++;
    return { allowed: true, reason: 'Rate limit check passed' };
  }

  private checkTimeWindow(rules: PolicyRules): { allowed: boolean; reason: string } {
    if (!rules.allowedHours) {
      return { allowed: true, reason: 'No time window restriction' };
    }

    const now = new Date();
    const hour = now.getHours();
    const { start, end } = rules.allowedHours;

    const inWindow = start <= end
      ? hour >= start && hour < end
      : hour >= start || hour < end; // wraps midnight

    if (!inWindow) {
      return { allowed: false, reason: `Current hour ${hour} outside allowed window ${start}-${end}` };
    }

    if (rules.allowedDays) {
      const day = now.getDay();
      if (!rules.allowedDays.includes(day)) {
        return { allowed: false, reason: `Day ${day} not in allowed days` };
      }
    }

    return { allowed: true, reason: 'Time window check passed' };
  }

  private checkStepUp(
    rules: PolicyRules,
    request: AccessRequest,
  ): { allowed: boolean; reason: string; pendingApprovalId?: string } {
    if (!rules.requireApproval) {
      return { allowed: true, reason: 'No step-up required' };
    }

    // Create a pending approval entry so the caller can wait for user decision.
    const approvalId = `stepup_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;

    return {
      allowed: false,
      reason: 'Step-up approval required',
      pendingApprovalId: approvalId,
    };
  }

  /**
   * Request step-up approval asynchronously. Returns a promise that resolves
   * when the user approves or rejects the request. Times out after the
   * specified duration (default: 60s).
   */
  requestApproval(
    approvalId: string,
    request: AccessRequest,
    timeoutMs = 60_000,
  ): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      const pending: PendingApproval = {
        id: approvalId,
        agentId: request.agentId,
        action: request.action,
        itemType: request.itemType,
        itemId: request.itemId,
        reason: 'Step-up approval required for this action',
        createdAt: Date.now(),
        resolve: (approved: boolean) => {
          this.pendingApprovals.delete(approvalId);
          resolve(approved);
        },
      };

      this.pendingApprovals.set(approvalId, pending);

      // Notify the bridge (extension popup) about the approval request
      this.onApprovalNeeded?.(pending);

      // Auto-reject on timeout
      setTimeout(() => {
        if (this.pendingApprovals.has(approvalId)) {
          this.pendingApprovals.delete(approvalId);
          resolve(false);
        }
      }, timeoutMs);
    });
  }

  /**
   * Resolve a pending approval (called when user clicks approve/deny in the extension).
   */
  resolveApproval(approvalId: string, approved: boolean): boolean {
    const pending = this.pendingApprovals.get(approvalId);
    if (!pending) return false;

    pending.resolve(approved);
    return true;
  }

  /** Get all pending approvals (for UI display). */
  getPendingApprovals(): PendingApproval[] {
    return Array.from(this.pendingApprovals.values());
  }

  resetCounters(): void {
    this.rateLimitCounters.clear();
  }
}

File: apps/macos/AuthBoxMac/Core/Policy/PolicyEngine.swift
//
//  PolicyEngine.swift
//  P4 — deny-by-default policy engine. Faithful port of
//  packages/mcp-protocol/src/policy-engine.ts.
//
//  Contract (fail-closed): no policies → deny; all disabled → deny; unknown
//  type → deny; any single failing policy → deny. An intent is allowed only
//  when EVERY enabled policy passes.
//
//  The clock is injectable so rate-limit and time-window logic is deterministic
//  under test (no wall-clock dependence).
//

import Foundation

struct PendingApproval: Identifiable, Equatable {
    let id: String
    let agentId: String
    let action: AgentAction
    let itemType: String?
    let itemId: String?
    let reason: String
    let createdAt: Date
}

@MainActor
final class PolicyEngine {
    private var rateLimitCounters: [String: (count: Int, windowStart: Date)] = [:]
    private var pendingApprovals: [String: PendingApproval] = [:]
    private var resolvers: [String: CheckedContinuation<Bool, Never>] = [:]
    private var timeoutTasks: [String: Task<Void, Never>] = [:]

    private let now: () -> Date
    private let makeApprovalId: () -> String

    /// Notified when a step-up approval is needed (the consent UI subscribes).
    var onApprovalNeeded: ((PendingApproval) -> Void)?

    init(now: @escaping () -> Date = Date.init,
         makeApprovalId: @escaping () -> String = { "stepup_\(UUID().uuidString)" }) {
        self.now = now
        self.makeApprovalId = makeApprovalId
    }

    // MARK: - Evaluation (deny-by-default)

    func evaluate(_ policies: [AgentPolicy], intent: AccessIntent) -> AccessEffect {
        guard !policies.isEmpty else {
            return AccessEffect(allowed: false, reason: "No policies defined", appliedPolicies: [])
        }
        let enabled = policies.filter(\.enabled).sorted { $0.priority > $1.priority }
        guard !enabled.isEmpty else {
            return AccessEffect(allowed: false, reason: "All policies disabled", appliedPolicies: [])
        }

        var applied: [String] = []
        for policy in enabled {
            applied.append(policy.id)
            let result = evaluateOne(policy, intent: intent)
            if !result.allowed {
                return AccessEffect(allowed: false, reason: result.reason,
                                    appliedPolicies: applied,
                                    pendingApprovalId: result.pendingApprovalId)
            }
        }
        return AccessEffect(allowed: true, reason: "All policies passed", appliedPolicies: applied)
    }

    private func evaluateOne(_ policy: AgentPolicy, intent: AccessIntent)
        -> (allowed: Bool, reason: String, pendingApprovalId: String?) {
        switch policy.policyType {
        case .item_scope:   return checkItemScope(policy.rules, intent)
        case .action_perm:  return checkActionPermission(policy.rules, intent)
        case .rate_limit:   return checkRateLimit(intent.agentId, policy.rules)
        case .time_window:  return checkTimeWindow(policy.rules)
        case .step_up:      return checkStepUp(policy.rules)
        }
    }

    // MARK: - Per-type checks

    private func checkItemScope(_ r: PolicyRules, _ intent: AccessIntent)
        -> (Bool, String, String?) {
        if let denied = r.deniedItemIds, denied.contains(intent.itemId ?? "") {
            return (false, "Item explicitly denied", nil)
        }
        if r.allowedItemTypes != nil, intent.itemType == nil {
            return (false, "Missing item type for scoped policy", nil)
        }
        if let types = r.allowedItemTypes, let t = intent.itemType, !types.contains(t) {
            return (false, "Item type '\(t)' not in allowed types", nil)
        }
        if r.allowedItemIds != nil, intent.itemId == nil {
            return (false, "Missing item ID for scoped policy", nil)
        }
        if let ids = r.allowedItemIds, let i = intent.itemId, !ids.contains(i) {
            return (false, "Item ID not in allowed list", nil)
        }
        if r.allowedFolderIds != nil, intent.folderId == nil {
            return (false, "Missing folder ID for scoped policy", nil)
        }
        if let folders = r.allowedFolderIds, let f = intent.folderId, !folders.contains(f) {
            return (false, "Folder not in allowed list", nil)
        }
        return (true, "Item scope check passed", nil)
    }

    private func checkActionPermission(_ r: PolicyRules, _ intent: AccessIntent)
        -> (Bool, String, String?) {
        guard let allowed = r.allowedActions, !allowed.isEmpty else {
            return (true, "No action restrictions", nil)
        }
        if !allowed.contains(intent.action) {
            return (false, "Action '\(intent.action.rawValue)' not permitted", nil)
        }
        return (true, "Action permitted", nil)
    }

    private func checkRateLimit(_ agentId: String, _ r: PolicyRules) -> (Bool, String, String?) {
        guard let maxRequests = r.maxRequests, let windowSeconds = r.windowSeconds else {
            return (true, "No rate limit configured", nil)
        }
        let current = now()
        let window = TimeInterval(windowSeconds)
        if let counter = rateLimitCounters[agentId],
           current.timeIntervalSince(counter.windowStart) <= window {
            if counter.count >= maxRequests {
                return (false, "Rate limit exceeded: \(maxRequests) requests per \(windowSeconds)s", nil)
            }
            rateLimitCounters[agentId] = (counter.count + 1, counter.windowStart)
            return (true, "Rate limit check passed", nil)
        }
        rateLimitCounters[agentId] = (1, current)
        return (true, "Rate limit check passed", nil)
    }

    private func checkTimeWindow(_ r: PolicyRules) -> (Bool, String, String?) {
        guard let hours = r.allowedHours else { return (true, "No time window restriction", nil) }
        let cal = Calendar.current
        let comps = cal.dateComponents([.hour, .weekday], from: now())
        let hour = comps.hour ?? 0
        let inWindow = hours.start <= hours.end
            ? (hour >= hours.start && hour < hours.end)
            : (hour >= hours.start || hour < hours.end)   // wraps midnight
        if !inWindow {
            return (false, "Current hour \(hour) outside allowed window \(hours.start)-\(hours.end)", nil)
        }
        if let days = r.allowedDays {
            let day = (comps.weekday ?? 1) - 1   // Calendar weekday is 1=Sun; normalize to 0=Sun
            if !days.contains(day) {
                return (false, "Day \(day) not in allowed days", nil)
            }
        }
        return (true, "Time window check passed", nil)
    }

    private func checkStepUp(_ r: PolicyRules) -> (Bool, String, String?) {
        guard r.requireApproval == true else { return (true, "No step-up required", nil) }
        return (false, "Step-up approval required", makeApprovalId())
    }

    // MARK: - Step-up approval (async, deny on timeout)

    /// Suspend until the user approves/denies, or the timeout auto-rejects.
    /// Register-before-notify: the resolver is installed synchronously inside the
    /// continuation body BEFORE onApprovalNeeded fires, so a consumer that resolves
    /// immediately (fast UI, or a test) never races a not-yet-registered resolver.
    func requestApproval(_ approvalId: String, intent: AccessIntent,
                         timeoutSeconds: Double = 60) async -> Bool {
        await withCheckedContinuation { (c: CheckedContinuation<Bool, Never>) in
            resolvers[approvalId] = c
            let pending = PendingApproval(
                id: approvalId, agentId: intent.agentId, action: intent.action,
                itemType: intent.itemType, itemId: intent.itemId,
                reason: "Step-up approval required for this action", createdAt: now())
            pendingApprovals[approvalId] = pending
            timeoutTasks[approvalId] = Task { @MainActor [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(timeoutSeconds * 1_000_000_000))
                self?.resolveApproval(approvalId, approved: false)   // fail-closed on timeout
            }
            onApprovalNeeded?(pending)
        }
    }

    @discardableResult
    func resolveApproval(_ approvalId: String, approved: Bool) -> Bool {
        guard let c = resolvers.removeValue(forKey: approvalId) else { return false }
        pendingApprovals.removeValue(forKey: approvalId)
        timeoutTasks.removeValue(forKey: approvalId)?.cancel()
        c.resume(returning: approved)
        return true
    }

    func getPendingApprovals() -> [PendingApproval] { Array(pendingApprovals.values) }
    func resetCounters() { rateLimitCounters.removeAll() }
}
